honeyflare/__init__.py

```diff
@@ -1,2 +1,197 @@
-def process_bucket_object(bucket, object_name):
-    pass
+import datetime
+import gzip
+import json
+import os
+import random
+import re
+import sys
+import threading
+
+import libhoney
+from google.api_core.exceptions import PreconditionFailed
+
+from .exceptions import RetriableError
+from .locks import GCSLock
+from .urlshape import compile_pattern, urlshape
+from .version import __version__
+
+
+STATUS_CODE_RE = re.compile(r'"EdgeResponseStatus":\s?(\d{3})')
+
+
+def process_bucket_object(
+        bucket,
+        object_name,
+        honeycomb_dataset,
+        honeycomb_key,
+        patterns=None,
+        query_param_filter=None,
+        sampling_rate_by_status=None,
+        lock_bucket=None,
+        ):
+    '''
+    :param bucket: A `google.cloud.storage.bucket.Bucket` logs should be
+        downloaded from.
+    :param object_name: The name of the object in the bucket to download.
+    :param honeycomb_dataset: The name of the honeycomb dataset to write to.
+    :param honeycomb_key: The honeycomb API key.
+    :param patterns: A list of path patterns to match against.
+    :param query_param_filter: A set of query parameters to allow. If None, all
+        will be allowed. If empty, none.
+    :param sampling_rate_by_status: A dictionary mapping a status code to a
+        sampling rate. Ie {200: 10, 400: 1}. A specific match will be checked
+        first (ie {404: 10}), then the general class of code (ie 400 for a 404).
+    :param lock_bucket: If you want to use a dedicated bucket for holding locks
+        and completion status, pass it here. Otherwise the bucket that holds the
+        logs will be used (requires write access to that bucket).
+    '''
+    if sampling_rate_by_status is None:
+        sampling_rate_by_status = {}
+
+    if lock_bucket is None:
+        lock_bucket = bucket
+
+    compiled_patterns = [compile_pattern(p) for p in patterns or []]
+    libhoney_client = create_libhoney_client(honeycomb_key, honeycomb_dataset)
+
+    lock = GCSLock(lock_bucket, 'locks/%s' % object_name)
+    total_events = 0
+    with lock:
+        if is_already_processed(lock_bucket, object_name):
+            # We might have been retried due to a failure but another
+            # function succeeded in the meantime
+            return
+
+        local_path = download_file(bucket, object_name)
+
+        for sample_rate, entry in get_sampled_file_entries(local_path, sampling_rate_by_status):
+            event = libhoney_client.new_event()
+            event.sample_rate = sample_rate
+            enrich_entry(entry, compiled_patterns, query_param_filter)
+            event.add(entry)
+            event.created_at = datetime.datetime.utcfromtimestamp(entry['EdgeEndTimestamp']/1e9)
+            event.send_presampled()
```

Comment:
As we discussed not super familiar with events vs trace/spans but I'm guessing this is to not double sample (as they are already filtered at `get_sampled_file_entries`)

Reply:
Correct.